codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -250,79 +251,164 @@ private void writePendingBytes() throws Http2Exception {
      * @param connectionWindow The connection window this is available for use at this point in the tree.
      * @return An object summarizing the write and allocation results.
      */
-    private int allocateBytesForTree(Http2Stream parent, int connectionWindow) {
+    static int allocateBytesForTree(Http2Stream parent, int connectionWindow) throws Http2Exception {
         FlowState state = state(parent);
         if (state.streamableBytesForTree() <= 0) {
             return 0;
         }
-        int bytesAllocated = 0;
         // If the number of streamable bytes for this tree will fit in the connection window
         // then there is no need to prioritize the bytes...everyone sends what they have
         if (state.streamableBytesForTree() <= connectionWindow) {
-            for (Http2Stream child : parent.children()) {
-                state = state(child);
-                int bytesForChild = state.streamableBytes();
-
-                if (bytesForChild > 0 || state.hasFrame()) {
-                    state.allocate(bytesForChild);
-                    bytesAllocated += bytesForChild;
-                    connectionWindow -= bytesForChild;
+            SimpleChildFeeder childFeeder = new SimpleChildFeeder(connectionWindow);
+            parent.forEachChild(childFeeder);
+            return childFeeder.bytesAllocated;
+        }
+
+        ChildFeeder childFeeder = new ChildFeeder(parent, connectionWindow);
+        // Iterate once over all children of this parent and try to feed all the children.
+        parent.forEachChild(childFeeder);
+
+        // Now feed any remaining children that are still hungry until the connection
+        // window collapses.
+        childFeeder.feedHungryChildren();
+
+        return childFeeder.bytesAllocated;
+    }
+
+    /**
+     * A {@link Http2StreamVisitor} that performs the HTTP/2 priority algorithm to distribute the available connection
+     * window appropriately to the children of a given stream.
+     */
+    private static final class ChildFeeder implements Http2StreamVisitor {
+        final int maxSize;
+        int totalWeight;
+        int connectionWindow;
+        int nextTotalWeight;
+        int nextConnectionWindow;
+        int bytesAllocated;
+        Http2Stream[] stillHungry;
+        int nextTail;
+
+        ChildFeeder(Http2Stream parent, int connectionWindow) {
+            maxSize = parent.numChildren();
+            totalWeight = parent.totalChildWeights();
+            this.connectionWindow = connectionWindow;
+            this.nextConnectionWindow = connectionWindow;
+        }
+
+        @Override
+        public boolean visit(Http2Stream child) throws Http2Exception {
+            FlowState state = state(child);
+            int weight = child.weight();
+            double weightRatio = weight / (double) totalWeight;
```

Comment:
I think I remember extracting this out but it is only used in 1 place.  Can we inline it?